ksqldb-rest-app/src/main/java/io/confluent/ksql/api/impl/QueryEndpoint.java

```diff
@@ -133,19 +132,36 @@ public QueryPublisher createQueryPublisher(
         sql, properties, sessionVariables);
 
     if (statement.getStatement().isPullQuery()) {
-      return createPullQueryPublisher(
-          context, serviceContext, statement, pullQueryMetrics, workerExecutor,
-          metricsCallbackHolder);
+      final ImmutableAnalysis analysis = ksqlEngine
+          .analyzeQueryWithNoOutput(statement.getStatement(), statement.getStatementText());
+      return createTablePullQueryPublisher(
+          analysis,
+          context,
+          serviceContext,
+          statement,
+          pullQueryMetrics,
+          workerExecutor,
+          metricsCallbackHolder
+      );
     } else if (ScalablePushUtil.isScalablePushQuery(statement.getStatement(), ksqlEngine,
         ksqlConfig, properties)) {
-      return createScalablePushQueryPublisher(context, serviceContext, statement, workerExecutor,
-          requestProperties);
+      final ImmutableAnalysis analysis = ksqlEngine
+          .analyzeQueryWithNoOutput(statement.getStatement(), statement.getStatementText());
+      return createScalablePushQueryPublisher(
+          analysis,
+          context,
+          serviceContext,
+          statement,
+          workerExecutor,
+          requestProperties
+      );
     } else {
       return createPushQueryPublisher(context, serviceContext, statement, workerExecutor);
```

Comment:
do non-scalable push queries not benefit from having the analysis also done up front? I was thinking why not do the analysis outside of the if/else tree?

Reply:
IIRC, the analysis currently needs to be a little different for transient and persistent queries, so I just didn't mess with it for now.

Reply:
After digging back through the code, I can confirm that was the reason, and I made another attempt to unify the paths, but changing transient queries has too much splash damage.